File: src/java/com/sudokusolver/main/SudokuProcessorTest.java
package com.sudokusolver.main;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.io.StringReader;
import java.io.StringWriter;

import org.junit.jupiter.api.Test;

public class SudokuProcessorTest {

    /**
     * Test: solve an example sudoku and test solution
     */
    @Test
    void testSmallSudoku() {
        StringReader input = new StringReader("""
                4 _ _ 1
                _ 1 3 _
                _ 4 1 _
                1 _ _ 3""");
        StringWriter output = new StringWriter();

        boolean result = SudokuProcessor.processSudoku(input, output);

        assertTrue(result);
        assertEquals("""
                4 3 2 1\s
                2 1 3 4\s
                3 4 1 2\s
                1 2 4 3\s
                """, output.toString());
    }
}

File: src/go/main/SudokuProcessor_test.go
package main_test

import (
	"strings"
	"testing"

	"github.com/mleemansnl/sudoku-solver/src/go/main"
	"github.com/stretchr/testify/suite"
)

// Test Suite: SudokuProcessor
type SudokuProcessorTestSuite struct {
	suite.Suite
}

// Go run entry to testify test suite
func TestSSudokuProcessorTestSuite(t *testing.T) {
	suite.Run(t, new(SudokuProcessorTestSuite))
}

/**
 * Test: solve an example sudoku and test solution
 */
func (suite *SudokuProcessorTestSuite) TestSmallSudoku() {
	// Solve small example sudoku
	input := `4 _ _ 1
_ 1 3 _
_ 4 1 _
1 _ _ 3`
	var output strings.Builder

	result := main.ProcessSudoku(
		strings.NewReader(input),
		&output,
	)

	suite.True(result)
	suite.Equal(`4 3 2 1 
2 1 3 4 
3 4 1 2 
1 2 4 3 
`, output.String())
}
